ledger/distribution/src/main.rs

```diff
@@ -420,15 +409,83 @@ fn main() {
             }
 
             next_block_num += 1;
-
-            let state = StateData {
-                next_block: next_block_num,
-            };
-            let json_data = serde_json::to_string(&state).expect("failed serializing state data");
-            fs::write(&state_file_path, json_data).expect("failed writing state file");
         }
 
         // TODO: make this configurable
         std::thread::sleep(std::time::Duration::from_millis(10));
     }
 }
+
+fn first_block_to_handle(ledger_db: &LedgerDB, block_handler: &dyn BlockHandler) -> u64 {
+    let num_blocks = ledger_db
+        .num_blocks()
+        .expect("Failed to get the number of blocks from the ledger database");
+    let last_upload_block = (0..num_blocks)
+        .rev()
+        .find(|block_index| block_handler.block_exists(*block_index));
+    match last_upload_block {
+        Some(last_upload_block) => last_upload_block + 1,
+        None => 0,
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use mc_blockchain_types::BlockVersion;
+    use mc_common::logger::test_with_logger;
+    use mc_ledger_db::test_utils::{create_ledger, initialize_ledger};
+    use mc_transaction_core::AccountKey;
+    use mc_util_test_helper::{RngType, SeedableRng};
+    use std::path::Path;
+    use tempfile::TempDir;
+    use walkdir::WalkDir;
+
+    fn number_of_files_in_directory(directory: impl AsRef<Path>) -> u64 {
+        WalkDir::new(directory)
+            .into_iter()
+            .filter_map(Result::ok)
+            .filter(|e| e.file_type().is_file())
+            .count()
+            .try_into()
+            .unwrap()
+    }
+
+    #[test_with_logger]
+    fn finding_first_block_to_distribute(logger: Logger) {
+        let mut rng = RngType::from_seed([0u8; 32]);
+        let key = AccountKey::random(&mut rng);
+
+        // Note: Number of blocks is one more than the block index.
+        // block indices (0, 1, 2, 3, 4) => number of blocks = 5
+        let number_of_blocks = 5;
+        let mut ledger = create_ledger();
+        initialize_ledger(
+            BlockVersion::MAX,
+            &mut ledger,
+            number_of_blocks,
+            &key,
+            &mut rng,
+        );
+
+        let temp_dir = TempDir::new().unwrap();
+        let distribution_path = temp_dir.path();
+        let mut block_handler = LocalBlockWriter::new(distribution_path.into(), logger);
+
+        for expected_block_count in [0, 1, 3, 5] {
+            // Note: the `0` case won't write any blocks.
+            // This will always re-write the blocks, but that's fine for this test.
+            for block_index in 0..expected_block_count {
+                block_handler.write_single_block(&ledger.get_block_data(block_index).unwrap());
+            }
```

Comment:
I wanted to flush out the logic and ensure there wasn't an off by one for computing the start block.
While I wasn't able to test the `S3::block_exists()` implementation, I think this proves out the `first_block_to_handle()` logic.